### PR TITLE
[checkpoint] Log timings for checkpoint IO save and load

### PR DESCRIPTION
**Pull request by ananthsub**
# What does this PR do ?

Adds timings around checkpoint save/load IO calls within Nemo1 and Nemo2 Lightning strategies.

To test, sanity checked logs from existing checkpointing tests:
```
pytest -v --log-cli-level=INFO --capture=no -s tests/core/test_dist_ckpt.py::TestAsyncSave::test_async_save_produces_same_checkpoints_as_sync
```

**Collection**: nlp, llm

# Changelog 
Adds timings around checkpoint save/load IO calls within Nemo1 and Nemo2 Lightning strategies

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/lightning/pytorch/strategies/fsdp_strategy.py [hunk 1]
```diff
@@ -225,7 +227,12 @@ def save_checkpoint(
                 checkpoint['optimizer'] = get_optimizer_state_dict(self.model, self.optimizers)
                 pyt_to_mcore_state_dict(checkpoint['optimizer']['state'], prefix="optimizer.state.")
 
+        start_time = time.monotonic()
         self.checkpoint_io.save_checkpoint(checkpoint, filepath, storage_options=storage_options)
+        end_time = time.monotonic()
+        logging.info(
```
Comment (ashbhandare) [inline nemo/lightning/pytorch/strategies/fsdp_strategy.py:233]: Is there a way to fetch the rank here

File: nemo/lightning/pytorch/strategies/fsdp_strategy.py [hunk 2]
```diff
@@ -257,7 +264,12 @@ def load_checkpoint(self, checkpoint_path: str | Path) -> Dict[str, Any]:
             pyt_to_mcore_state_dict(osd['state'], prefix="optimizer.state.")
             sharded_state_dict["optimizer"] = osd
 
+        start_time = time.monotonic()
         checkpoint = self.checkpoint_io.load_checkpoint(path, sharded_state_dict=sharded_state_dict)
+        end_time = time.monotonic()
+        logging.info(
```
Comment (ashbhandare) [inline nemo/lightning/pytorch/strategies/fsdp_strategy.py:270]: same as above

File: nemo/lightning/pytorch/strategies/megatron_strategy.py [hunk 3]
```diff
@@ -749,7 +750,12 @@ def save_checkpoint(
             if self.ckpt_save_optimizer:
                 checkpoint["optimizer"] = [self.optimizer_sharded_state_dict()]
 
+        start_time = time.monotonic()
         self.checkpoint_io.save_checkpoint(checkpoint, filepath, storage_options=storage_options)
+        end_time = time.monotonic()
+        logging.info(
```
Comment (ashbhandare) [inline nemo/lightning/pytorch/strategies/megatron_strategy.py:756]: Same, adding rank

File: nemo/lightning/pytorch/strategies/megatron_strategy.py [hunk 4]
```diff
@@ -781,9 +787,15 @@ def load_checkpoint(self, checkpoint_path: Union[str, Path], selective_restore:
         strict = (
             self.lightning_module.strict_loading if self.ckpt_load_strictness is None else self.ckpt_load_strictness
         )
+
+        start_time = time.monotonic()
         checkpoint = self.checkpoint_io.load_checkpoint(
             checkpoint_path, sharded_state_dict=sharded_state_dict, strict=strict
         )
+        end_time = time.monotonic()
```
Comment (ashbhandare) [inline nemo/lightning/pytorch/strategies/megatron_strategy.py:795]: Same, adding rank

**Comment by ashors1** [comment]
Overall, changes LGTM. Just one question -- what is the intended behavior when async checkpointing is enabled? Do you think these logs will be misleading in that case?

**Comment by ananthsub** [comment]
>  what is the intended behavior when async checkpointing is enabled? Do you think these logs will be misleading in that case?

In this case we'd report the time spent blocked while staging the checkpoint. In follow ups I'd like to make the logging more granular so that we can report time spent on staging vs finalization for the async case

**Comment by ashors1** [comment]
Sorry, one other comment. Do you think we should consider making this logging optional? Or provide the option to log only a single rank? If we're running a job on many GPUs, I wonder if this would clutter the logs.

**Comment by maanug-nv** [comment]
What are your thoughts on using a decorator for this? I think one actually already exists [here](https://github.com/NVIDIA/NeMo/blob/main/nemo/utils/callbacks/dist_ckpt_io.py#L62-L68). If we need to start logging runtime of more functions in general, I think using a decorator prevents disrupting readability.

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 5]
```diff
@@ -420,7 +426,12 @@ def save_checkpoint(
             # PTL override to accomodate model parallel checkpoints
             filepath = inject_model_parallel_rank(filepath)
             if self.is_global_zero or app_state.data_parallel_rank == 0:
+                start_time = time.monotonic()
                 self.checkpoint_io.save_checkpoint(checkpoint, filepath, storage_options=storage_options)
+                end_time = time.monotonic()
```
Comment (maanug-nv) [inline nemo/collections/nlp/parts/nlp_overrides.py:431]: just wondering why here vs `dist_checkpointing.save()` inside `checkpoint_io.save_checkpoint()`?

**Comment by maanug-nv** [comment]
Just curious if the calls in these locations should be timed as well:
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/parts/nlp_overrides.py#L527
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/parts/nlp_overrides.py#L1371
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/parts/nlp_overrides.py#L1044

**Comment by maanug-nv** [comment]
> . Do you think we should consider making this logging optional? Or provide the option to log only a single rank? If we're running a job on many GPUs, I wonder if this would clutter the logs.

Or use `logging.debug`  level?

**Comment by apaithankar** [comment]
Could you also include the iteration number? Filename is optional.

Here is the doc for requirements from Heimdall: https://docs.google.com/document/d/1J5TVmg4msY1HA0kcR6etb4Ha72gBRC2aPen1Z0rinQI/edit?usp=sharing

**Review by maanug-nv — APPROVED** [review]
lgtm, thanks!
